boot/assets_test.go

```diff
@@ -0,0 +1,407 @@
+	d := c.MkDir()
+
+	tab := bootloadertest.Mock("trusted-assets", "").WithTrustedAssets()
+	// MockBootloader does not implement trusted assets
+	bootloader.Force(tab)
+	defer bootloader.Force(nil)
+	tab.TrustedAssetsList = []string{
+		"asset",
+		"nested/other-asset",
+	}
+
+	// we get an observer for UC20
+	uc20Model := makeMockUC20Model()
+	obs, err := boot.TrustedAssetsInstallObserverForModel(uc20Model, d)
+	c.Assert(obs, NotNil)
+	c.Assert(err, IsNil)
+
+	data := []byte("foobar")
+	// SHA3-384
+	dataHash := "0fa8abfbdaf924ad307b74dd2ed183b9a4a398891a2f6bac8fd2db7041b77f068580f9c6c66f699b496c2da1cbcc7ed8"
+	err = ioutil.WriteFile(filepath.Join(d, "foobar"), data, 0644)
+	c.Assert(err, IsNil)
+
+	_, err = obs.Observe(gadget.ContentWrite, mockRunBootStruct, boot.InitramfsUbuntuBootDir,
+		filepath.Join(d, "foobar"), "asset")
+	c.Assert(err, IsNil)
+	// observe same asset again
+	_, err = obs.Observe(gadget.ContentWrite, mockRunBootStruct, boot.InitramfsUbuntuBootDir,
+		filepath.Join(d, "foobar"), "asset")
+	c.Assert(err, IsNil)
+	// different one
+	_, err = obs.Observe(gadget.ContentWrite, mockRunBootStruct, boot.InitramfsUbuntuBootDir,
+		filepath.Join(d, "foobar"), "nested/other-asset")
+	c.Assert(err, IsNil)
+	// a non trusted asset
+	_, err = obs.Observe(gadget.ContentWrite, mockRunBootStruct, boot.InitramfsUbuntuBootDir,
+		filepath.Join(d, "foobar"), "non-trusted")
+	c.Assert(err, IsNil)
+	// a single file in cache
+	checkContentGlob(c, filepath.Join(dirs.SnapBootAssetsDir, "trusted-assets", "*"), []string{
+		filepath.Join(dirs.SnapBootAssetsDir, "trusted-assets", fmt.Sprintf("asset-%s", dataHash)),
+		filepath.Join(dirs.SnapBootAssetsDir, "trusted-assets", fmt.Sprintf("other-asset-%s", dataHash)),
+	})
+	// the list of trusted assets was asked for just once
+	c.Check(tab.TrustedAssetsCalls, Equals, 1)
+	// let's see what the observer has tracked
+	tracked := obs.CurrentTrustedBootAssetsMap()
+	c.Check(tracked, DeepEquals, boot.BootAssetsMap{
+		"asset":       []string{dataHash},
+		"other-asset": []string{dataHash},
+	})
+}
+
+func (s *assetsSuite) TestInstallObserverNonTrustedBootloader(c *C) {
+	d := c.MkDir()
+
+	// MockBootloader does not implement trusted assets
+	bootloader.Force(bootloadertest.Mock("mock", ""))
+	defer bootloader.Force(nil)
+
+	// we get an observer for UC20
+	uc20Model := makeMockUC20Model()
+	obs, err := boot.TrustedAssetsInstallObserverForModel(uc20Model, d)
+	c.Assert(obs, NotNil)
+	c.Assert(err, IsNil)
+
+	err = ioutil.WriteFile(filepath.Join(d, "foobar"), []byte("foobar"), 0644)
+	c.Assert(err, IsNil)
+	// bootloder is found, but ignored because it does not support trusted assets
+	_, err = obs.Observe(gadget.ContentWrite, mockRunBootStruct, boot.InitramfsUbuntuBootDir,
+		filepath.Join(d, "foobar"), "asset")
+	c.Assert(err, IsNil)
+	c.Check(osutil.IsDirectory(dirs.SnapBootAssetsDir), Equals, false,
+		Commentf("%q exists while it should not", dirs.SnapBootAssetsDir))
```

Comment:
should we check also that obs.CurrentTrustedBootAssetsMap() returns something empty?